src/main/java/uk/gov/justice/hmpps/datacompliance/services/OffenderRetentionService.java

```diff
@@ -2,44 +2,106 @@
-import uk.gov.justice.hmpps.datacompliance.web.dto.ManualRetentionReason;
-import uk.gov.justice.hmpps.datacompliance.web.dto.ManualRetentionReasonCode;
+import uk.gov.justice.hmpps.datacompliance.repository.jpa.repository.RetentionReasonCodeRepository;
+import uk.gov.justice.hmpps.datacompliance.security.UserSecurityUtils;
+import uk.gov.justice.hmpps.datacompliance.utils.TimeSource;
+import uk.gov.justice.hmpps.datacompliance.web.dto.ManualRetentionRequest;
 
+import javax.annotation.Nullable;
+import javax.persistence.OptimisticLockException;
+import javax.transaction.Transactional;
+import java.util.Objects;
 import java.util.Optional;
 
-import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
 
 @Service
+@Transactional
 @AllArgsConstructor
 public class OffenderRetentionService {
 
+    private final TimeSource timeSource;
+    private final UserSecurityUtils userSecurityUtils;
     private final ManualRetentionRepository manualRetentionRepository;
+    private final RetentionReasonCodeRepository retentionReasonCodeRepository;
 
     public Optional<ManualRetention> findManualOffenderRetention(final OffenderNumber offenderNumber) {
-        return manualRetentionRepository.findFirstByOffenderNoOrderByRetentionDateTimeDesc(offenderNumber.getOffenderNumber())
-                .map(this::transform);
+        return manualRetentionRepository.findFirstByOffenderNoOrderByRetentionVersionDesc(offenderNumber.getOffenderNumber());
     }
 
-    private ManualRetention transform(
-            final uk.gov.justice.hmpps.datacompliance.repository.jpa.model.retention.ManualRetention manualRetention) {
+    /**
+     * Updates / creates a retention record for the offender.
+     *
+     * Uses optimistic locking to ensure the client is updating the latest version of the record.
+     *
+     * @param ifMatchValue Uses the 'If-Match' header (for an update) to check the client has the latest version of the record.
+     * @return Returns the previous version of the record if an update, or empty if a creation.
+     */
+    public Optional<ManualRetention> updateManualOffenderRetention(final OffenderNumber offenderNumber,
+                                                                   final ManualRetentionRequest manualRetentionRequest,
+                                                                   @Nullable final String ifMatchValue) {
 
-        return ManualRetention.builder()
-                .offenderNo(manualRetention.getOffenderNo())
-                .modifiedDateTime(manualRetention.getRetentionDateTime())
-                .staffId(manualRetention.getStaffId())
-                .retentionReasons(manualRetention.getManualRetentionReasons().stream()
-                        .map(this::transform)
-                        .collect(toList()))
+        final var existingRecord = findManualOffenderRetention(offenderNumber);
```

Comment:
Just going by the name of the method it may help to include previous in `existingRecord` var???